File: codewars/authoring/java/src/codewars/_ISO.java
package codewars;

import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Stream;
import java.util.List;
import java.util.stream.Collectors;

/*
 * so, when are two type, `a` and `b`, considered equal?
 * a definition might be, it is possible to go from `a` to `b`,
 * and from `b` to `a`.
 * Going a roundway trip should leave you the same value.
 * Unfortunately it is virtually impossible to test this in Haskell.
 * This is called ISO.
 * <p>
 * ISO can also be simulated in Java.
 * <p>
 * Created by ice1000 on 17-6-19.
 */
public abstract class ISO<A, B> {
		/* go backward */
		abstract A bw(final B b);

		/* go forward */
		abstract B fw(final A a);

		abstract static class Either<L, R> {
				public abstract <T> T pm(Function<L, T> lt, Function<R, T> rt);
    
				static <L, R> Either<L, R> left(L l) {
						return new Either<L, R>() {
								@Override
								public <T> T pm(Function<L, T> lt, Function<R, T> rt) {
										return lt.apply(l);
								}
        
								@Override
								public boolean equals(Object rhs) {
										return ((Either<L, R>) rhs).<Boolean>pm(l::equals, rr -> false);
								}
        
								@Override
								public String toString() {
										return "left: " + l.toString();
								}
						};
				}

				static <L, R> Either<L, R> right(R r) {
						return new Either<L, R>() {
								@Override
								public <T> T pm(Function<L, T> lt, Function<R, T> rt) {
										return rt.apply(r);
								}
        
								@Override
								public boolean equals(Object rhs) {
										return ((Either<L, R>) rhs).<Boolean>pm(ll -> false, r::equals);
								}

								@Override
								public String toString() {
										return "right: " + r.toString();
								}
						};
				}
		}

		final static class Tuple<A, B> {
				A a;
				B b;

				Tuple(A a, B b) {
						this.a = a;
						this.b = b;
				}
    
				@Override
				public boolean equals(Object rhs) {
						Tuple<A, B> ab = (Tuple<A, B>) rhs;
						return a.equals(ab.a) && b.equals(ab.b);
				}
		}

		final static class Unit {
				public static Unit INSTANCE = new Unit();

				private Unit() {
				}
		}

		static abstract class Void {
				public abstract <T> T absurd();
		}

		private static <A, B> ISO<A, B> iso(Function<A, B> forward, Function<B, A> backward) {
				return new ISO<A, B>() {
						@Override public A bw(B b) { return backward.apply(b); }
						@Override public B fw(A a) { return forward.apply(a); }
				};
		}

		/* given ISO a b, we can go from a to b */
		static <A, B> Function<A, B> subStL(final ISO<A, B> iso) {
				return iso::fw;
		}

		/* and vise versa */
		static <A, B> Function<B, A> subStR(final ISO<A, B> iso) {
				return iso::bw;
		}

		/* There can be more than one ISO a b */
		static ISO<Boolean, Boolean> isoBool() {
				return refl();
		}

		static ISO<Boolean, Boolean> isoBoolNot() {
				return iso(a -> !a, b -> !b);
		}

		/* isomorphism is reflexive */
		static <A> ISO<A, A> refl() {
				return iso(a -> a, b -> b);
		}

		/* isomorphism is symmetric */
		static <A, B> ISO<A, B> symm(final ISO<B, A> iso) {
				return iso(iso::bw, iso::fw);
		}

		/* isomorphism is transitive */
		static <A, B, C> ISO<A, C> trans(final ISO<A, B> ab, final ISO<B, C> bc) {
				return iso(a -> bc.fw(ab.fw(a)), c -> ab.bw(bc.bw(c)));
		}

		/* We can combine isomorphism */
		static <A, B, C, D> ISO<Tuple<A, C>, Tuple<B, D>> isoTuple
				(final ISO<A, B> ab, final ISO<C, D> cd) {
				return iso(ac -> new Tuple<>(ab.fw(ac.a), cd.fw(ac.b)),
									 bd -> new Tuple<>(ab.bw(bd.a), cd.bw(bd.b)));
		}

		static <A, B> ISO<Stream<A>, Stream<B>> isoStream(final ISO<A, B> iso) {
				return iso(as -> as.map(iso::fw), bs -> bs.map(iso::bw));
		}

		static <A, B> ISO<Optional<A>, Optional<B>> isoOptional(final ISO<A, B> iso) {
				return iso(a -> a.map(iso::fw), b -> b.map(iso::bw));
		}

		static <A, B, C, D> ISO<Either<A, C>, Either<B, D>> isoEither
				(final ISO<A, B> ab, final ISO<C, D> cd) {
				return iso(l -> l.pm(a -> Either.left(ab.fw(a)),
														 c -> Either.right(cd.fw(c))),
									 r -> r.pm(b -> Either.left(ab.bw(b)),
														 d -> Either.right(cd.bw(d))));
		}

		/*
		 * Going another way is hard (and is generally impossible)
		 * Remember, for all valid ISO, converting and converting back
		 * is the same as the original value.
		 * You need this to prove some case are impossible.
		 */
		static <A, B> ISO<A, B> isoUnOptional(final ISO<Optional<A>, Optional<B>> iso) {
				return iso(a -> iso.fw(Optional.of(a))
									 .orElseGet(() -> iso.fw(Optional.empty()).get()),
									 b -> iso.bw(Optional.of(b))
									 .orElseGet(() -> iso.bw(Optional.empty()).get()));
		}

		// We cannot have
		// isoUnEither ::
		// ISO (Either a b) (Either c d) -> ISO a c -> ISO b d.
		// Note that we have
		static ISO<Either<Stream<Unit>, Unit>, Either<Stream<Unit>, Void>>isoEU() {
				return iso(l -> l.pm(ll -> Either.left(Stream.concat(Stream.of(Unit.INSTANCE), ll)),
														 r -> Either.left(Stream.empty())),
									 r -> r.pm(l -> { 
													 List<Unit> lu = l.collect(Collectors.toList());
													 return lu.isEmpty() ? 
															 Either.right(Unit.INSTANCE) :
															 Either.left(Stream.generate
																					 (() -> Unit.INSTANCE).limit(lu.size() - 1));
											 },
											 Void::absurd));
    }

		// where (), the empty tuple, has 1 value, and Void has 0 value
		// If we have isoUnEither,
		// We have ISO () Void by calling isoUnEither isoEU
		// That is impossible, since we can get a Void by
		// substL on ISO () Void
		// So it is impossible to have isoUnEither

		static <A, B, C, D> ISO<Function<A, C>, Function<B, D>>
				isoFunc(final ISO<A, B> ab, final ISO<C, D> cd) {
				return iso(ac -> a -> cd.fw(ac.apply(ab.bw(a))),
									 bd -> b -> cd.bw(bd.apply(ab.fw(b))));
		}

		/* And we have isomorphism on isomorphism! */
		static <A, B> ISO<ISO<A, B>, ISO<B, A>> isoSymm() {
				return iso(ISO::symm, ISO::symm);
		}
}

File: codewars/authoring/kotlin/ISO.kt
import java.util.Optional
import java.util.stream.Stream
import java.util.stream.Collectors.*
import java.util.stream.Stream.*

/*
 * so, when are two type, `a` and `b`, considered equal?
 * a definition might be, it is possible to go from `a` to `b`,
 * and from `b` to `a`.
 * Going a roundway trip should leave you the same value.
 * Unfortunately it is virtually impossible to test this in Haskell.
 * This is called ISO.
 * <p>
 * ISO can also be simulated in Java.
 * <p>
 * Created by ice1000 on 17-6-19.
 */
abstract class ISO<A, B> {
	/** go backward */
	abstract fun bw(b: B): A

	/** go forward */
	abstract fun fw(a: A): B
}

abstract class Either<out L, out R> {
	abstract fun <T> pm(lt: (L) -> T, rt: (R) -> T): T
}

fun <L, R> left(l: L): Either<L, R> = object : Either<L, R>() {
	override fun <T> pm(lt: (L) -> T, rt: (R) -> T): T = lt(l)
	override fun equals(other: Any?) = (other as Either<*, *>).pm({ l == it }, { false })
	override fun toString() = "left: $l"
}

fun <L, R> right(r: R): Either<L, R> = object : Either<L, R>() {
	override fun <T> pm(lt: (L) -> T, rt: (R) -> T): T = rt(r)
	override fun equals(other: Any?) = (other as Either<*, *>).pm({ false }, { r == it })
	override fun toString() = "right: $r"
}

private fun <T> Nothing.absurd(): T = this

private inline fun <A, B> iso(crossinline forward: (A) -> B, crossinline backward: (B) -> A): ISO<A, B> = object : ISO<A, B>() {
	override fun bw(b: B): A = backward(b)
	override fun fw(a: A): B = forward(a)
}

/* given ISO a b, we can go from a to b */
fun <A, B> subStL(iso: ISO<A, B>): (A) -> B = iso::fw

/* and vise versa */
fun <A, B> subStR(iso: ISO<A, B>): (B) -> A = iso::bw

/* There can be more than one ISO a b */
fun isoBool(): ISO<Boolean, Boolean> = refl()

fun isoBoolNot(): ISO<Boolean, Boolean> = iso(Boolean::not, Boolean::not)

/* isomorphism is reflexive */
fun <A> refl(): ISO<A, A> = iso({ it }, { it })

/* isomorphism is symmetric */
fun <A, B> symm(iso: ISO<B, A>): ISO<A, B> = iso(iso::bw, iso::fw)

/* isomorphism is transitive */
fun <A, B, C> trans(ab: ISO<A, B>, bc: ISO<B, C>): ISO<A, C> = iso({ bc.fw(ab.fw(it)) }, { ab.bw(bc.bw(it)) })

/* We can combine isomorphism */
fun <A, B, C, D> isoPair(ab: ISO<A, B>, cd: ISO<C, D>): ISO<Pair<A, C>, Pair<B, D>> =
		iso({ (a, c) -> ab.fw(a) to cd.fw(c) }, { (b, d) -> ab.bw(b) to cd.bw(d) })

fun <A, B> isoStream(iso: ISO<A, B>): ISO<Stream<A>, Stream<B>> = iso({ it.map(iso::fw) }, { it.map(iso::bw) })
fun <A, B> isoOptional(iso: ISO<A, B>): ISO<Optional<A>, Optional<B>> = iso({ it.map(iso::fw) }, { it.map(iso::bw) })

fun <A, B, C, D> isoEither(ab: ISO<A, B>, cd: ISO<C, D>): ISO<Either<A, C>, Either<B, D>> =
		iso({ it.pm({ left(ab.fw(it)) }, { right(cd.fw(it)) }) }, { it.pm({ left(ab.bw(it)) }, { right(cd.bw(it)) }) })

/*
* Going another way is hard (and is generally impossible)
* Remember, for all valid ISO, converting and converting back
* is the same as the original value.
* You need this to prove some case are impossible.
*/
fun <A, B> isoUnOptional(iso: ISO<Optional<A>, Optional<B>>): ISO<A, B> = iso({ iso.fw(Optional.of(it)).orElseGet { iso.fw(Optional.empty()).get() } },
		{ iso.bw(Optional.of(it)).orElseGet { iso.bw(Optional.empty()).get() } })

// We cannot have
// isoUnEither ::
// ISO (Either a b) (Either c d) -> ISO a c -> ISO b d.
// Note that we have
fun isoEU(): ISO<Either<Stream<Unit>, Unit>, Either<Stream<Unit>, Nothing>> = iso({ it.pm({ left(concat(of(Unit), it)) }, { left(empty()) }) },
		{ it.pm({ it.collect(toList()).let { if (it.isEmpty()) right(Unit) else left(generate { Unit }.limit(it.size - 1L)) } }, { it.absurd() }) })

// where (), the empty Pair, has 1 value, and Void has 0 value
// If we have isoUnEither,
// We have ISO () Void by calling isoUnEither isoEU
// That is impossible, since we can get a Void by
// substL on ISO () Void
// So it is impossible to have isoUnEither

fun <A, B, C, D> isoFunc(ab: ISO<A, B>, cd: ISO<C, D>): ISO<(A) -> C, (B) -> D> =
		iso({ ac -> { cd.fw(ac(ab.bw(it))) } },
				{ bd -> { cd.bw(bd(ab.fw(it))) } })

/* And we have isomorphism on isomorphism! */
fun <A, B> isoSymm(): ISO<ISO<A, B>, ISO<B, A>> = iso(::symm, ::symm)